sound/soc/sof/sof-audio.c

```diff
@@ -256,29 +256,128 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
 	return 0;
 }
 
+/*
+ * free all widgets in the sink path starting from the source widget
+ * (DAI type for capture, AIF type for playback)
+ */
+static int sof_free_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget *widget,
+				    int dir)
+{
+	struct snd_soc_dapm_path *p;
+	int err;
+	int ret = 0;
+
+	/* free all widgets even in case of error to keep use counts balanced */
+	snd_soc_dapm_widget_for_each_sink_path(widget, p) {
+		if (p->sink->dobj.private && widget->dobj.private) {
+			if (WIDGET_IS_AIF(widget->id) || WIDGET_IS_DAI(widget->id)) {
+				err = sof_widget_free(sdev, widget->dobj.private);
+				if (err < 0)
+					ret = err;
+			}
+
+			err = sof_widget_free(sdev, p->sink->dobj.private);
+			if (err < 0)
+				ret = err;
+
+			err = sof_free_widgets_in_path(sdev, p->sink, dir);
```

Comment:
what happens if there is a loop in the dapm graph?

Reply:
good point. let me add a condition to stop at the starting widget in case of a loop.

Reply:
the loop could be just after the starting widget... you may need something that tracks all widgets to see if you've already handled them.

Wondering if there isn't already defined in ASoC/DAPM provides for 'safe' walks through a graph with loops?

Reply:
@plbossart It was quite simple to prevent an infinite walk. we can mark the paths as walking or not as we traverse the list.